include/pybind11/cast.h

```diff
@@ -34,6 +34,20 @@ template <typename Type> class opaque {
 
 NAMESPACE_BEGIN(detail)
 
+/// Convert from any type pointer to another avoding warnings about void* vs (void*()) incompatibility.
```

Comment:
I agree with your assessment that it's too much code for something that doesn't do anything. And "templated function returning pointer to function" syntax is extremely unusual and very unreadable.

Unfortunately I have not been able to find GCC switch that disables this particular warning. And all stack overflow answers I seen suggest casting, although they typically offer this solution:

` *(void**)(&f)` or something along these lines.